### PR TITLE
ACL2

### LINKED ISSUE #171: New ACL
Current ACL rules are annoying to use, create something better.

### PR DESCRIPTION
**Pull request by sinus-x**
There is a lot going on in this PR. But I've split it into commits as I have worked on it, so it should be easier to dig through.

Resolves #171

### CODE REVIEW

**Comment by sinus-x** [comment]
This version fully deprecates the old ACL check. Maybe we should be more graceful, and allow the old database tables and `check.acl` work further, until all external repositories are updated, too.

**Review by czechbol — APPROVED** [review]
I think this is looking good, just we should update our other modules too. We'll see how things go from there

File: modules/base/acl/module.py [hunk 1]
```diff
@@ -25,494 +23,606 @@ def __init__(self, bot):
     #
```
Comment (ArcasCZ) [inline modules/base/acl/module.py:23]: What does this lonely hash do here? 🤔

File: modules/base/acl/module.py [hunk 2]
```diff
@@ -25,494 +23,606 @@ def __init__(self, bot):
...
 
-        To set up virtual group with no link to Discord roles, set ``role_id``
-        to ``0``.
-        """
-        RE_NAME = r"[a-zA-Z-]+"
-        if re.fullmatch(RE_NAME, name) is None:
+    @check.acl2(check.ACLevel.GUILD_OWNER)
+    @acl_mapping_.command(name="add")
+    async def acl_mapping_add(self, ctx, role: nextcord.Role, level: str):
+        """Add ACL level to role mappings."""
```
Comment (ArcasCZ) [inline modules/base/acl/module.py:72]: Is it possible to give user more information about those levels?

File: modules/base/acl/module.py [hunk 3]
```diff
@@ -25,494 +23,606 @@ def __init__(self, bot):
...
+        await ctx.reply(
+            _(ctx, "Role **{role}** will be mapped to **{level}**.").format(
+                role=role.name, level=level.name
+            )
+        )
+        await guild_log.info(
             ctx.author,
             ctx.channel,
-            f"New ACL group '{name}' (parent '{parent}', role ID '{role_id}').",
+            f"New ACLevel mapping from role '{role.name}' to level '{level.name}'.",
```
Comment (ArcasCZ) [inline modules/base/acl/module.py:96]: I think this is less confusing for pepople not knowing this function.

"New ACLevel mapping for role '{role.name}' set to level '{level.name}'."

File: modules/base/acl/module.py [hunk 4]
```diff
@@ -25,494 +23,606 @@ def __init__(self, bot):
...
+        )
+
+        for page in table:
+            await ctx.send("```" + page + "```")
+
+    @check.acl2(check.ACLevel.GUILD_OWNER)
+    @acl_default_.command("add")
+    async def acl_default_add(self, ctx, command: str, level: str):
+        """Add custom ACLevel for a command."""
+        # TODO When we know how to access the decorators, check that
```
Comment (ArcasCZ) [inline modules/base/acl/module.py:155]: I leave this command here and try to figure out how to do this.

File: modules/base/acl/module.py [hunk 5]
```diff
@@ -25,494 +23,606 @@ def __init__(self, bot):
...
+            ctx.author,
+            ctx.channel,
+            f"ACLevel mapping for role '{role.name}' removed.",
+        )
 
-        ``param`` -- ``value`` options:
+    @check.acl2(check.ACLevel.SUBMOD)
+    @acl_.group(name="default")
+    async def acl_default_(self, ctx):
+        """Manage ACLevel defaults of commands."""
```
Comment (ArcasCZ) [inline modules/base/acl/module.py:118]: What are defaults good for? I'm, as a person, who sees this first, kinda confused, without looking for docs.

**Review by ArcasCZ — COMMENTED** [review]
It looks great. I think this solves all ACL problems we have. Only one thing I'm missing is manual ACL checking in case of some interactions (buttons and so on).

File: modules/base/acl/module.py (hunk 4 above)
Comment (ArcasCZ) [reply modules/base/acl/module.py:155]: What about this?

```Python
    @debug_.command(name="acl")
    async def debug_acl(self, ctx: commands.Context):
        command = self.bot.get_command("module reload").callback
        
        source = inspect.getsource(command)
        index = source.find("def ")
        decs = [
            line.strip().split()[0]
            for line in source[:index].strip().splitlines()
            if line.strip()[0] == "@"
        ]
        
        print(decs)
```

File: modules/base/acl/module.py (hunk 5 above)
Comment (sinus-x) [reply modules/base/acl/module.py:118]: I'm still not sure about the command names. This is used to overwrite hardcoded defaults:
```py
@check.acl2(check.ACLevel.MEMBER)
async def foo(self, ctx):
    ...
```
can be updated by calling `!acl default set foo EVERYONE`. Then you don't need to have any role to use that command, if that's what you want/need on your server.

ACL2 provides sane defaults, but guild owners should have the option to change the defaults.

File: modules/base/acl/module.py (hunk 2 above)
Comment (sinus-x) [reply modules/base/acl/module.py:72]: There will be documentation on this topic. But we may do something to make it easier.

File: modules/base/acl/module.py (hunk 1 above)
Comment (sinus-x) [reply modules/base/acl/module.py:23]: It's visual separator between sections. It's not necessary, but I tend to use them from time to time, especially in bigger modules.

File: modules/base/acl/module.py (hunk 5 above)
Comment (ArcasCZ) [reply modules/base/acl/module.py:118]: Oh, I see. In this case I think that this should be said in short version in this group description. Something like:

```Python
"""Manage ACL default level assignment for commands."""```

Something like that?

File: modules/base/acl/module.py (hunk 2 above)
Comment (ArcasCZ) [reply modules/base/acl/module.py:72]: I think just simple list of levels or command to list levels would be good enough.

File: modules/base/acl/module.py (hunk 1 above)
Comment (ArcasCZ) [reply modules/base/acl/module.py:23]: I see, I try to do the same. Maybe just add label to it? Like ` # HELPER FUNCTIONS` or ` # LISTENERS`

**Pull request merged**